Mathlib/Order/Booleanisation.lean

```diff
@@ -0,0 +1,263 @@
+/-
+Copyright (c) 2023 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.Order.Hom.Lattice
+
+/-!
+# Adding complements to a generalized Boolean algebra
+
+This file embeds any generalized Boolean algebra into a Boolean algebra.
+
+This concretely proves that any equation holding true in the theory of Boolean algebras that does
+not reference `ᶜ` also holds true in the theory of generalized Boolean algebras. Put another way,
+one does not need the existence of complements to prove something which does not talk about
+complements.
+
+## Main declarations
+
+* `Booleanisation`: Boolean algebra containing a given generalised Boolean algebra as a sublattice.
+* `Booleanisation.inlLatticeHom`: Boolean algebra containing a given generalised Boolean algebra as
+  a sublattice.
+-/
+
+open Function
+
+variable {α : Type*}
+
+/-- Boolean algebra containing a given generalised Boolean algebra `α` as a sublattice.
+
+This should be thought of as made of a copy of `α` (representing elements of `α`) living under
+another copy of `α` (representing complements of elements of `α`). -/
+def Booleanisation (α : Type*) := α ⊕ α
+
+namespace Booleanisation
+
+instance instDecidableEq [DecidableEq α] : DecidableEq (Booleanisation α) := Sum.instDecidableEqSum
+
+variable [GeneralizedBooleanAlgebra α] {x y : Booleanisation α} {a b : α}
+
+/-- The natural inclusion `a ↦ a` from a generalized Boolean algebra to its generated Boolean
+algebra. -/
+@[match_pattern] abbrev inl : α → Booleanisation α := Sum.inl
+
+/-- The inclusion `a ↦ aᶜ from a generalized Boolean algebra to its generated Boolean algebra. -/
+@[match_pattern] abbrev inr : α → Booleanisation α := Sum.inr
+
+/-- The order on `Booleanisation α` is as follows: For `a b : α`,
+* `a ≤ b` iff `a ≤ b` in `α`
+* `a ≤ bᶜ` iff `a` and `b` are disjoint in `α`
+* `aᶜ ≤ bᶜ` iff `b ≤ a` in `α`
+* `¬ aᶜ ≤ b` -/
+protected inductive LE : Booleanisation α → Booleanisation α → Prop
+  | protected inl {a b} : a ≤ b → Booleanisation.LE (inl a) (inl b)
+  | protected inr {a b} : a ≤ b → Booleanisation.LE (inr b) (inr a)
+  | protected sep {a b} : Disjoint a b → Booleanisation.LE (inl a) (inr b)
+
+/-- The order on `Booleanisation α` is as follows: For `a b : α`,
+* `a < b` iff `a < b` in `α`
+* `a < bᶜ` iff `a` and `b` are disjoint in `α`
+* `aᶜ < bᶜ` iff `b < a` in `α`
+* `¬ aᶜ < b` -/
+protected inductive LT : Booleanisation α → Booleanisation α → Prop
+  | protected inl {a b} : a < b → Booleanisation.LT (inl a) (inl b)
+  | protected inr {a b} : a < b → Booleanisation.LT (inr b) (inr a)
+  | protected sep {a b} : Disjoint a b → Booleanisation.LT (inl a) (inr b)
+
+instance instLE : LE (Booleanisation α) where
+  le := Booleanisation.LE
+
+instance instLT : LT (Booleanisation α) where
+  lt := Booleanisation.LT
```

Comment:
Maybe use `@[inherit_doc]` here to copy over the documentation from the definition of `LE` and `LT`?